### ISSUE TITLE
Remove execution # of cells pasted between notebooks

### ISSUE THREAD

**Issue by jlumpe** [jupyterlab/jupyterlab#11570]
### Problem

When copying and pasting one or more cells from one notebook to another, the execution # (e.g. `In [123]`) gets copied as well. The number doesn't have any meaning in the new notebook and can be kind of confusing if you're relying on it to know which cells have been run since the last restart.

### Proposed Solution

Pasted cell execution # should be cleared if destination notebook is different than source. Output should probably still be kept.

**Comment by schmidi314** [comment]
Also: The execution # isn't removed after closing and re-opening the notebook, with kernel also restarted. Is this a bug or a feature? :) I'd find it more natural if all cells that haven't been executed in the current session have a clean "In [ ]".

**Comment by krassowski** [comment]
Do you mean to clear the execution count each time after opening the notebook? I don't think this beneficial because execution count holds information which facilitates reprodicibility. Out-of-order execution is possible in Jupyter notebooks, and having the information about it stored is a way of safeguarding from users mis-interpreting why outputs look like they look.
But maybe I did not understand your suggestion.

**Comment by schmidi314** [comment]
Yes, that's what I meant. Just a few days ago I had a situation with a not-so-jupyter-familiar colleague who got confused about what was executed already during I showed him something over skype.

I get your point on the reproducibility, though. Maybe we could get the best of both worlds by playing with the colors? Like the classic notebook blue for "executed in current session" and the grey for the previous execs.

**Comment by krassowski** [comment]
Colours are hard (can be distracting/non colorblind-friendly). We could make old results fainter though.

My first thought was something link broken link icon (🔗 but with a breakage) but that is visually complex and takes space.
Maybe in place of the brackets we could use a variation of the brackets to indicate that the result is from a previous session? Where we currently have: `[ 1 ]`

we could try to also use: `⟦ 1 ⟧`, `⧘ 1 ⧙`, `⦋ 1 ⦌`, `⁅ 1 ⁆`

or something more prominent. I don't know if this is the right path forward. Something tells me that probably we should design a custom set of SVGs for different states rather than try to re-use a bracket variation (but it does not hurt to experiment with brackets).

**Comment by schmidi314** [comment]
Yes, I like those ideas. I just prepared some veeery preliminary quick'n'dirty branch on my fork to play with the visuals. Output area not touched, kernel restart not touched. I put curly brackets there and played with the colors. With the different brackets I dislike `{  }` for new cells. I could keep the square brackets for fresh cells?

https://mybinder.org/v2/gh/schmidi314/jupyterlab/scratch/execution_count?urlpath=lab-dev

But I noticed that at the moment we can't figure out if a pasted cell came originally from the same notebook or not. The Clipboard from lumino doesn't support that. So I guess addressing the original issue would require to extend the clipboard, right?

Another thing: What about the pre-previous-session cells? We should really clear those to avoid confusion.

**Comment by krassowski** [comment]
I was thinking about a way of marking cells executed in previous session (so for which the variables are not say "live" in the kernel), and not about copy-paste. I would say that if we copy within the same notebook we could assume that cell (output) is "live" and when we copy to another notebook it is not (let's call it "stale" or "potentially outdated"), so copy-paste becomes a special case of marking cells which were not executed in the current session as "stale".

> Another thing: What about the pre-previous-session cells? We should really clear those to avoid confusion.

Personal opinion: no, I would prefer to keep them (maybe mark them somehow); making change like that would be breaking the experience for many many users who see it as a feature or rely on this behaviour, and that would be a destructive (in the sense that it removes information) change.

**Comment by schmidi314** [comment]
Hm. Ok, are you thinking about something like `{5; 3}`, meaning the fifth execution in the (pre)^3-vious session?

**Comment by krassowski** [comment]
UX-wise my hunch is that it would be too much information to show the "3" on the screen, but it certainly could be shown on hover via `title` attribute. This is an idea that is worth exploring but as you note the implementation would require leaving more trace/metadata about execution in the notebook and some API changes. It might even benefit from a Jupyter Enhancement Proposal (or it might not, I am not sure yet).

**Comment by schmidi314** [comment]
So we have a number of independent issues there: (1) where to store the information about "Session previosity" [what would be a good word for that?] (2) how to display the information (3) how to manage cut/copy/paste

Regarding (1) I just stored it in the cell metadata for the moment.

Regarding (2): I like the double wavy brackets best. Single wavy has not the right width. Curly brackets seem misleading to me.
![previosity](https://user-images.githubusercontent.com/74381262/144764373-523b22c2-d66d-4db9-ad3a-fcd70207158f.gif)
: 

Regarding (3): Postpone


https://mybinder.org/v2/gh/schmidi314/jupyterlab/feature/session_previosity?urlpath=lab-dev

**Comment by davidbrochart** [comment]
I'm wondering if there is not an overlap with the dirty indicator:
![image](https://user-images.githubusercontent.com/4711805/144766019-abcc71ff-c22a-45bb-a4d9-a967ac3c0611.png)
The dot basically means that the code in the input cell is not the one that was executed to produce the output. But it could be replaced with wavy brackets, with a message on hover saying "not executed in current session", what do you think?

**Comment by jlumpe** [comment]
I didn't realize the dirty indicator was a thing until just now, but that definitely looks like a good way to do it (maybe with a different symbol and/or color).

"Not executed in current session" seems like a good flag for both situations (before last restart in current notebook and pasted in from another notebook), I'm not sure it would be necessary to further differentiate between these two.

**Comment by schmidi314** [comment]
@davidbrochart great point and yes, I also see some overlap.

Here's my interpretation of the present situation:

1) **dirty**: output (potentially) doesn't match code + kernel state (potentially) doesn't match code.
2) **"not executed in present session"**: kernel state (certainly) doesn't match code [no statement about output]

I would view the information about when the code was last executed to be able to reproduce what happened as subordinate to the second point.

But I don't see how to make them more orthogonal or even unify them without losing information. For instance, if we have a new session (no execs yet) and I start to change a code cell that already has some output. `not_exec` will always be false then and I need extra information to figure out if the output still matches the input, right?
 
Note that `dirty` disappears when I revert changes, which is a really neat feature; it even distinguishes some irrelevant changes like spaces and new lines. Whoever did this: great job :)

**Comment by davidbrochart** [comment]
> But I don't see how to make them more orthogonal or even unify them without losing information. For instance, if we have a new session (no execs yet) and I start to change a code cell that already has some output. `not_exec` will always be false then and I need extra information to figure out if the output still matches the input, right?

Right, looks like we need `dirty` **and** `not_exec`.

> Note that `dirty` disappears when I revert changes, which is a really neat feature; it even distinguishes some irrelevant changes like spaces and new lines. Whoever did this: great job :)

It's @martinRenou, maybe he has some ideas here.

**Comment by martinRenou** [comment]
> Note that dirty disappears when I revert changes, which is a really neat feature; it even distinguishes some irrelevant changes like spaces and new lines. Whoever did this: great job :)

Thanks :D To be fair, it's pretty much everything it does, it could be smarter by looking into the code AST, but that would be more work on the implementation and language dependent.

> (1) where to store the information about "Session previosity"

I see in the GIF that you show if it was 2 sessions before or in the previous session. Is this information that important?

I wonder if it would be sensible to always reset the inputs execution count to `null` when starting a new kernel, while keeping the outputs?
This would be enough to convey the idea that the cell has not been executed yet but was executed in a previous session.
